### ISSUE TITLE
Discrete values generated from DCM (bug?)

### ISSUE THREAD

**Issue by laurenyatlas** [bachlab/PsPM#99]
### Summary
I am measuring SCR epochs of ~10s when heat is administered, or during heat anticipation. We see multiphasic responses, so SF seems most appropriate. I've compared the AUC approach with DCM, and while I get nice continuous responses in the sf.stats field from AUC, the DCM approach generates only a few specific values for all trials. I noticed this with an earlier version of PsPM as well. Is this a feature or a bug? 

Across 168 trials, there are only 12 different output values in the sf.stats field from the DCM approach:

       0
    0.0826
    0.1235
    0.1250
    0.1653
    0.2469
    0.2479
    0.3306
    0.3704
    0.3750
    0.4132
    0.4938
    0.4959

This is quite different from AUC. You can also see the repetition of these specific values in the figures I'm attaching.
![example_SCR_PSPM_AUCvsDCM_8554run1](https://user-images.githubusercontent.com/26441167/76557406-e5573080-6471-11ea-957e-d85d4ce32ed2.png)
![example_SCR_PSPM_AUCvsDCM_heat_allruns](https://user-images.githubusercontent.com/26441167/76557422-ec7e3e80-6471-11ea-9464-e1adae221153.png)

Here's a zipped mat file with the job and the output.
[SF_DCM_job_and_output.zip](https://github.com/bachlab/PsPM/files/4325963/SF_DCM_job_and_output.zip)

### Technical Info
* PsPM revision ID: 4.2.1 (also saw in earlier versions)
* MATLAB version: 2018B

**Comment by irojkov-ph** [comment]
Hi @laurenyatlas , 
First of all, thank you for this issue !
Unfortunately, it is difficult for me to find out what exactly is happening in `pspm_dcm` using only the files you have sent me, so may I ask you to send me your file called `tpspm_physio_for_pspm2_run1.mat`. In case you do not want to disclose your data publicly you can always send them via email: ivan.rojkov@uzh.ch .
As soon as I have the data I'll do my best to answer your question !
Ivan.

**Comment by laurenyatlas** [comment]
No problem, here you go! 
[tpspm_physio_for_pspm2_run1.mat.zip](https://github.com/bachlab/PsPM/files/4329629/tpspm_physio_for_pspm2_run1.mat.zip)

**Comment by irojkov-ph** [comment]
Thank you very much for that !
Unfortunately, I have some difficulties to reproduce your analysis.
The timing matrix I've extracted from your first zip files seems to give me errors because it is wrongly written, but probably it is not the right matrix I've extracted :/ Can you please provide me the timing matrix or the model file you have used for dcm ?
Best, 
Ivan

**Comment by laurenyatlas** [comment]
Hi Ivan,
Strange. The mat file with “job” in the title should run using the pspm_jobman to completely replicate the DCM job. Does that not work?
Lauren

**Comment by irojkov-ph** [comment]
I wanted to do it manually instead of using `pspm_jobman` because it is easier to debug but I think I made a mistake somewhere. So I abandoned this idea and used that function. I reproduced both of your results (AUC and DCM) and here is my conclusion:

The most important thing to remember is that `sf.stats` are intrinsically different quantities for DCM and AUC ! Indeed:
 - For AUC, it is basically `mean(scr-min(scr))` where `scr` is your signal for a given timing interval and it corresponds to the area under the curve of an SCR time series.
 - For DCM, `sf.stats` corresponds to the frequency (Hz) of responses above threshold (0.1 in your case). And the responses are the posterior mean of the evolution parameters (4 parameters in total per time series). The formula is `numel(find(responses>threshold))/(numel(scr)/sr)`.

What I will do now is to discuss that with Dominik to see if it is indeed what `sf.stats` should return and come back to you as soon as I can.

Sorry for the slow progress (current world's situation is not helpful).

**Comment by laurenyatlas** [comment]
I am grateful for any help, and obviously this is not as urgent as other matters! I recognize that the values have different meanings, but still think it is very strange that the values for dcm would be a discrete set, basically. Is there a different output field that would be more useful, or is this what you think captures the appropriate response and recommend people use for analysis?

Sent from my iPhone

On Mar 18, 2020, at 4:41 AM, irojkov-ph <notifications@github.com> wrote:

﻿

I wanted to do it manually instead of using pspm_jobman because it is easier to debug but I think I made a mistake somewhere. So I abandoned this idea and used that function. I reproduced both of your results (AUC and DCM) and here is my conclusion:

The most important thing to remember is that sf.stats are intrinsically different quantities for DCM and AUC ! Indeed:

  *   For AUC, it is basically mean(scr-min(scr)) where scr is your signal for a given timing interval and it corresponds to the area under the curve of an SCR time series.
  *   For DCM, sf.stats corresponds to the frequency (Hz) of responses above threshold (0.1 in your case). And the responses are the posterior mean of the evolution parameters (4 parameters in total per time series). The formula is numel(find(responses>threshold))/(numel(scr)/sr).

What I will do now is to discuss that with Dominik to see if it is indeed what sf.stats should return and come back to you as soon as I can.

Sorry for the slow progress (current world's situation is not helpful).

**Comment by dominikbach** [comment]
Hi Lauren!

as Ivan says, sf.stats is the number of above-threshold responses per second. This is an often-used quantity in research on tonic arousal, e.g. in the context of anxiety. The traditional way of deriving is by manually counting the peaks. 

PsPM emulates that method by estimating the input time series (similar to Ledalab, but with a more appropriate response function and a physiological sparsity constraint of no more than 1 SN burst in 2 seconds). It then counts the number of peaks and divides that integer number by the duration of the duration (to account for variable epoch duration). In your case, because the intervals are relatively short and many have the same duration, and because you cannot have more than 1/2 responses per second, there is only a small number of output values you can possibly get.

This quantity (as well as the manual way of doing it) predicts experimentally induced anxiety better than the amplitude of responses, or a combined measure of number and amplitude (you can see that if you compare Bach, Friston, Dolan 2010 and Bach, Daunizeau et al. 2011)). Notably, AUC is just number x amplitude in a LTI system (see for experimental validation Bach, Friston, Dolan 2010).

To summarise: the discrete number of values derives from the fact that you are estimating an integer quantity and this is a feature, not a bug. Whether this is appropriate in the context of heat pain, I don't know. It may also be worth checking whether the threshold of 0.1 µS is appropriate.

Finally, you can access the estimated input time series in the field  sf.model{1}.inv. This gives you, for every epoch, a vector of estimated time points (they can be negative due to the lag of the SN/SCR transmission) and amplitudes of the responses, their number, estimated time series yhat, and some housekeeping and troubleshooting information.

Hope this helps!
Best
Dominik
